### PR TITLE
Fix unstable formatting involving unwrapping multiple parentheses (#836)

### PR DESCRIPTION
**Pull request by Jma353**
This is a fix for #836.  To stabilize the formatting, I introduce a function which dedupes adjacent invisible parentheses, so that downstream formatting logic involving breaking up lines depending on line length get the same input regardless of whether `((({atom})))` or `({atom})` is attempting to be formatted.

Open to feedback on this solution and additional test cases I can throw at it; tried to be conservative w.r.t. how these parens are deduped.

As an aside, perhaps in a future PR we can refactor how `maybe_make_parens_invisible_in_atom` looks.  Currently, it's responsible for returning a boolean *and* causing side-effects, which isn't obvious off the bat.  I added a comment to clarify this for now.

### CODE REVIEW

File: black.py [hunk 1]
```diff
@@ -2872,7 +2872,15 @@ def normalize_invisible_parens(node: Node, parens_after: Set[str]) -> None:
...
             if child.type == syms.atom:
-                if maybe_make_parens_invisible_in_atom(child, parent=node):
+                # Determines if the underlying atom should be surrounded with
+                # invisible params - also makes parens invisible recursively
+                # within the atom
+                should_surround_with_parens = maybe_make_parens_invisible_in_atom(
+                    child, parent=node
+                )
+                # Removes repeated invisible parens within in the atom
+                dedupe_invisible_atom_parens(child)
```
Comment (JelleZijlstra) [inline black.py]: Is there a way we could do maybe_make_parens_invisible_in_atom() and dedupe_invisible_atom_parens() in one go, so we don't have to recurse over the expression twice?

**Comment by JelleZijlstra** [comment]
Also, thanks for doing this! It's great to get these unstable formatting issues sorted out.

File: black.py (hunk 1 above)
Comment (Jma353) [reply black.py]: Yea I can try and tweak this a bit.  Thanks for feedback on the approach.

**Comment by Jma353** [comment]
Addressed your comments @JelleZijlstra and added some more tests.  Should be ready for a second round of review.

**Comment by Jma353** [comment]
Ping @JelleZijlstra for review.

**Comment by zsol** [comment]
lgtm, thanks for working on this

**Pull request merged**